mapstructure.go

```diff
@@ -453,6 +453,13 @@ func (d *Decoder) decode(name string, input interface{}, outVal reflect.Value) e
 		return nil
 	}
 
+	if outVal.CanAddr() {
+		v := outVal.Addr()
+		if u, ok := v.Interface().(decoder); ok {
+			return u.DecodeMapstructure(input)
```

Comment:
I think it might make more sense to put this after the `DecodeHook`, so that any custom preprocessing can still happen.